ssl/quic/quic_route.c

```diff
@@ -0,0 +1,326 @@
+    if (r == NULL) {
+        OPENSSL_free(r);
+        return NULL;
+    }
+    ossl_list_remote_ids_init(&r->remote_id_list);
+    return r;
+}
+
+void ossl_quic_route_table_free(QUIC_ROUTE_TABLE *routes)
+{
+    QUIC_ROUTE *p, *pnext;
+
+    if (routes != NULL) {
+        lh_QUIC_ROUTE_free(routes->remote_id_tbl);
+        for (p = ossl_list_remote_ids_head(&routes->remote_id_list);
+             p != NULL; p = pnext) {
+            pnext = ossl_list_remote_ids_next(p);
+            OPENSSL_free(p);
+        }
+        OPENSSL_free(routes);
+    }
+}
+
+QUIC_ROUTE *ossl_route_table_get0_route_from_local(
+        ossl_unused const QUIC_ROUTE_TABLE *cache,
+        const unsigned char *conn_id, size_t conn_id_len)
+{
+    QUIC_ROUTE *r, conn;
+
+    if (!ossl_quic_conn_id_set(&conn.local, conn_id, conn_id_len))
+        return NULL;
+    if (!local_connection_read_lock())
+        return NULL;
+    r = lh_QUIC_ROUTE_retrieve(local_id_tbl, &conn);
+    local_connection_unlock();
+    return r;
+}
+
+QUIC_ROUTE *ossl_route_table_get0_route_from_remote(
+        const QUIC_ROUTE_TABLE *cache,
+        const unsigned char *conn_id, size_t conn_id_len)
+{
+    QUIC_ROUTE *r, conn;
+
+    if (!ossl_quic_conn_id_set(&conn.remote, conn_id, conn_id_len))
+        return NULL;
+    r = lh_QUIC_ROUTE_retrieve(cache->remote_id_tbl, &conn);
+    return r;
+}
+
+QUIC_ROUTE *ossl_quic_route_new(QUIC_ROUTE_TABLE *routes,
+                                const unsigned char *remote_id, size_t remote_id_len)
+{
+    QUIC_ROUTE *r = OPENSSL_malloc(sizeof(*r));
+    QUIC_ROUTE *p;
+
+    if (r != NULL) {
+        r->retired = 0;
+        r->tbl = routes;
+        if (!ossl_quic_conn_id_set(&r->remote, remote_id, remote_id_len)
+                || !local_connection_write_lock())
+            goto err_no_unlock;
+        p = ossl_list_remote_ids_head(&routes->remote_id_list);
+        r->seq_no = p != NULL ? p->seq_no + 1 : 0;
+        ossl_quic_conn_id_generate(&r->local);
+
+        if (lh_QUIC_ROUTE_insert(local_id_tbl, r) != NULL
+                || lh_QUIC_ROUTE_error(local_id_tbl) > 0)
+            goto err;
+        if (lh_QUIC_ROUTE_insert(routes->remote_id_tbl, r) != NULL
+                || lh_QUIC_ROUTE_error(routes->remote_id_tbl) > 0) {
+            lh_QUIC_ROUTE_delete(local_id_tbl, r);
+            goto err;
+        }
+        ossl_list_remote_ids_init_elem(r);
+        ossl_list_remote_ids_insert_head(&routes->remote_id_list, r);
+        local_connection_unlock();
+    }
+    return r;
+
+ err:
+    local_connection_unlock();
+ err_no_unlock:
+    OPENSSL_free(r);
+    return NULL;
+}
+
+/*
+ * Retire by sequence number up to and including the one specified.
+ */
+int ossl_quic_route_retire(QUIC_ROUTE_TABLE *routes, uint64_t seq_no)
+{
+    QUIC_ROUTE *p;
+
+    for (p = ossl_list_remote_ids_tail(&routes->remote_id_list);
+         p != NULL && p->seq_no <= seq_no;
+         p = ossl_list_remote_ids_prev(p))
+        p->retired = 1;
```

Comment:
No lock held for this....is this going to be a problem? Will tsan complain about this?